src/OrbitGl/ThreadTrack.cpp

```diff
@@ -501,8 +506,19 @@ void ThreadTrack::UpdatePrimitives(Batcher* batcher, uint64_t min_tick, uint64_t
         batcher->AddVerticalLine(pos, box_height, draw_data.z, color, std::move(user_data));
       }
 
-      // Use the time at boundary of the next pixel as a threshold to avoid overdraw.
-      next_pixel_start_time_ns = GetNextPixelBoundaryTimeNs(pos[0] + size[0], draw_data);
+      // Optimization when going through the scopes, as it doesn't have any sense to draw two timers
+      // in the same pixel.
+      // TODO (b/199736246): NextPixelBoundaryTimeNs should be greater than the pixel provided.
+      next_pixel_start_time_ns =
+          std::max(timer_info.end() + 1, GetNextPixelBoundaryTimeNs(pos[0] + size[0], draw_data));
+
+      node_to_draw = ordered_nodes.upper_bound(next_pixel_start_time_ns);
```

Comment:
Shouldn't we use `lower_bound` here? We want the first scope that starts at the time that is not less than `next_pixel_start_time_ns`, i.e. greater or equal to `next_pixel_start_time_ns`, which we'd get from using `lower_bound`, no? If that's the case, then we can remove the block below.

Reply:
We used to use lower_bound here, but I don't think that it's correct. Lower_bound gave you the greater or equal, but we don't want that. We want the **lower or equal**. 

Why? Because we are looking by start_time, but there is sometimes a case where start_time is before our query and end_time is after. In that case, we should anyway draw the timer. So, we are looking for the lower or equal to check if the end is after that time. If this is the case we will draw that timer, if not, we skip it.

How to calculate the lower or equal? is Upper_bound-1 (with the exception of upper_bound being the first node)

Reply:
ah yes, that makes sense, thanks.